### PR TITLE
ddl/failtest: migrate test-infra to testify

### LINKED ISSUE #27669: migrate test-infra to testify for `ddl/failtest` pkg

### PR DESCRIPTION
**Pull request by karuppiah7890**
### What problem does this PR solve?

Issue Number: close #27669 

Problem Summary:

### What is changed and how it works?

What's Changed: Migrated tests to use testify package instead of github.com/pingcap/check package for assertions and any test utils and matchers that github.com/pingcap/check provided

There are some small changes in `ddl/ddl_test` too

Note:
- When I ran `make gotest` I noticed that two tests are timing out. I need help with figuring out why that's the case and if it's happening in others machines and in CI too or just my machine. The two tests are - `TestT/TestAddIndexFailed` and `TestT/TestAddIndexWorkerNum`. Log -

<details>
   <summary>Logs (click to expand)</summary>

```bash
tidb $ make gotest
Running in native mode.
timeout-check
grep 'PASS:' gotest.log | go run tools/check/check-timeout.go || { $(find $PWD/ -type d | grep -vE "(\.git|tools)" | xargs tools/bin/failpoint-ctl disable); exit 1; }
The following test cases take too long to finish:
    --- PASS: TestT/TestAddIndexFailed (71.08s)
    --- PASS: TestT/TestAddIndexWorkerNum (55.20s)
PASS: builtin_time_vec_generated_test.go:6727: testVectorizeSuite1.TestVectorizedBuiltinTimeEvalOneVecGenerated	8.240s
PASS: builtin_time_vec_test.go:575: testVectorizeSuite2.TestVectorizedBuiltinTimeEvalOneVec	9.063s
--- PASS: TestClusterConfigInfoschema (6.09s)
--- PASS: TestCTEPreviewAndReport (6.78s)
exit status 252
make: *** [gotest] Error 1
tidb $ 
```

</details>

- I put lot of commits. It will be easier to review I guess. But I can squash it too once all review comments are done. Either squash it to one commit or a few commits but currently there are 24 commits!
- I added all the tests to run in a sequential manner, assuming `SerialSuites` of github.com/pingcap/check helps to run tests in series. Also, when I tried it in parallel with `t.Parallel()` there were errors which I didn't understand
- I added some `TODO`s with some questions, please do check them
- What is the purpose of this piece of code? -
```go
colsStr := ""
idxsStr := ""
for _, col := range cols {
	colsStr += col.Name.L + " "
}
for _, idx := range tbl.Meta().Indices {
	idxsStr += idx.Name.L + " "
}
```
`colsStr` and `idxsStr` don't seem to be used anywhere else, so I'm curious if it's dead code or if some test code got removed by someone by mistake

Tests

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by karuppiah7890** [comment]
cc @tisonkun could you please review or tag someone / some folks you know who can help with reviewing this?

**Comment by karuppiah7890** [comment]
Just now noticed the failure in Jenkins CI - idc-jenkins-ci-tidb/check_dev - https://ci.pingcap.net/job/tidb_ghpr_check/23317/display/redirect. I'll check it out and get back 👍

**Comment by karuppiah7890** [comment]
Looks like the failure is because I have migrated to testify. I need to probably change the name of `testFailDBSuite` to something in such a way that `./tools/check/check_testSuite.sh` doesn't think that test suite is not enabled

**Comment by karuppiah7890** [comment]
I also noticed another CI failure just now - https://ci.pingcap.net/job/tidb_ghpr_check_2/30519/display/redirect . I'll check it out

**Comment by karuppiah7890** [comment]
Ah okay,  idc-jenkins-ci-tidb/check_dev_2 - https://ci.pingcap.net/job/tidb_ghpr_check_2/30519/display/redirect has two failures. I recognize the second one - the one I mentioned about timeout of two tests. Looks like it's clear that the two tests have become slower than 5 seconds unlike before, something to checkout

<details>
  <summary>Logs(click to expand)</summary>

```bash
[2021-09-01T18:17:49.442Z] ++ grep module go.mod

[2021-09-01T18:17:49.442Z] ++ head -n 1

[2021-09-01T18:17:49.442Z] ++ awk '{print $2}'

[2021-09-01T18:17:49.442Z] + package_base=github.com/pingcap/tidb

[2021-09-01T18:17:49.442Z] + sed -i 's,go list ./...| grep -vE "cmd",go list ./...| grep -vE "cmd" | grep -vE "store/tikv$$",' ./Makefile

[2021-09-01T18:17:49.442Z] + '[' master == master ']'

[2021-09-01T18:17:49.442Z] + EXTRA_TEST_ARGS='-timeout 9m'

[2021-09-01T18:17:49.442Z] + make test_part_2

[2021-09-01T18:23:26.087Z] exit status 252

[2021-09-01T18:23:26.087Z] make: *** [gotest] Error 1

[2021-09-01T18:23:26.087Z] + cat test.log

[2021-09-01T18:23:26.087Z] cd tools/check; \

[2021-09-01T18:23:26.087Z] GO111MODULE=on go build -o ../bin/failpoint-ctl github.com/pingcap/failpoint/failpoint-ctl

[2021-09-01T18:23:26.087Z] Running in native mode.

[2021-09-01T18:23:26.087Z] timeout-check

[2021-09-01T18:23:26.087Z] grep 'PASS:' gotest.log | go run tools/check/check-timeout.go || { $(find $PWD/ -type d | grep -vE "(\.git|tools)" | xargs tools/bin/failpoint-ctl disable); exit 1; }

[2021-09-01T18:23:26.087Z] The following test cases take too long to finish:

[2021-09-01T18:23:26.087Z]     --- PASS: TestT/TestAddIndexFailed (43.90s)

[2021-09-01T18:23:26.087Z]     --- PASS: TestT/TestAddIndexWorkerNum (55.33s)

[2021-09-01T18:23:26.087Z] + cat test.log

[2021-09-01T18:23:26.087Z] + grep -Ev '^\[[[:digit:]]{4}(/[[:digit:]]{2}){2}'

[2021-09-01T18:23:26.087Z] + grep -A 30 '\-------'

[2021-09-01T18:23:26.087Z] + grep -A 29 FAIL

script returned exit code 1
```
</details>

**Comment by karuppiah7890** [comment]
TODO
- Fix `make check` failure - `make testSuite` failure specifically
- Fix timeout failure for `TestT/TestAddIndexFailed` test - should finish within 5 seconds
- Fix timeout failure for `TestT/TestAddIndexWorkerNum` test - should finish within 5 seconds

**Comment by tisonkun** [comment]
> Fix timeout failure for TestT/TestAddIndexFailed test - should finish within 5 seconds
> Fix timeout failure for TestT/TestAddIndexWorkerNum test - should finish within 5 seconds

Please check whether or not it is caused by your changes. Our CI env is a bit ( if not quite :( ) overloaded, for timeout cases you can often retrigger by `/run-check_dev_2`

**Comment by tisonkun** [comment]
Please avoid using `suite` as mentioned https://github.com/pingcap/tidb/pull/26082#issuecomment-877230972 .

You can take a look at how we do for `domain` and `planner/cascade` for example.

**Comment by tisonkun** [comment]
Also please avoid `TestT` pattern which is a dirty hack for go-check to work.

**Comment by tisonkun** [comment]
See also https://github.com/pingcap/tidb/pull/27277

**Comment by karuppiah7890** [comment]
I actually had some questions around the patterns and running tests parallely and of course using separate golang test functions. I was just wondering about the global config and didn't have context on it. But I can check it out and get back. It shouldn't be a big change I guess. Also, I wasn't using any testify suite, I have never used it before actually and have no idea about it, so no worries there

I'll get back with my thoughts. Looks like there's a communication gap since I didn't put all my thoughts about the PR. But yeah, I still don't have much context on the tests themselves except for a few things, so I guess that's what is hindering me. I'll reach out with specific questions, mostly about "global config", setup, tear down related. I'll first read some code and docs and then get back I guess, and try to answer some of these questions and many other questions (specific to each test) myself by checking code and discussions

**Comment by karuppiah7890** [comment]
TODO:
- [ ] Assign reviewer(s) once the above review comments are taken care of and previous TODO items are done

**Comment by tisonkun** [comment]
@karuppiah7890 this issue, after a closer look, is too complex to a first-time contributor. I push a commit to adopt all of our best practice on migrating tests, including:

* using `main_test.go` for global setup logic, and migrate testleak to goleak
* avoid TestT and clean up unnecessary code
* migrate CommonHandleSuite
* ...

you can take a look at the diff and give feedback. I may make some mistake also.

**Comment by karuppiah7890** [comment]
Ah. I did see similar `TestMain` in other places. And I don't have much context, so I don't have much to say. The only thing I was planning on avoiding was [merging from master to branch](https://github.com/pingcap/tidb/pull/27747/commits/eb7ad94424db5c0269b83a24b49862cac90e4460) - instead I usually rebase with latest master and force push in order to avoid merge commits which I find weird usually

**Comment by karuppiah7890** [comment]
If you think this is too complex, you can take it up 👍

**Comment by tisonkun** [comment]
> avoiding was merging from master to branch

Never mind. We always use squash and merge so that one PR results in one commit on master.

**Comment by karuppiah7890** [comment]
I'll probably look at other issues or just use TiDB for now and come back later to contribute 👍

**Comment by karuppiah7890** [comment]
I have removed my assignment for now, assuming you are continuing with this. And thanks for all the help! I was looking forward to contributing to database projects and especially something complex as I was always looking at good-first-issues. This PR was kind of interesting, I got to know what kind of tests databases do, which kind of seemed obvious in some cases, and I'm going to learn about failpoints which I'm curious about and seems to be used all over in this project's tests

**Comment by tisonkun** [comment]
@karuppiah7890 you're welcome. You're encouraged to ask any question or share ideas about design or impl on our [developer discussion forum](https://internals.tidb.io/).

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by karuppiah7890** [comment]
When commits are squashed automatically, what's the commit author of the final commit? Is is only the latest commit's author?

**Comment by tisonkun** [comment]
> When commits are squashed automatically, what's the commit author of the final commit? Is is only the latest commit's author?

@karuppiah7890 it's the PR author.

**Comment by tisonkun** [comment]
> Looks like the branch is out of date. Should I update it? With `rebase`

TiDB uses a "merge master" strategy which force keep up-to-date before PR merging and thus running tests with all commits.

However, it is not a strong requirement to keep up-to-date during reviewing since almost changes don't have conflicts. We sometimes click "Update branch" to keep up-to-date but after one merge commit, rebase is hard to perform.

Short answer, you shouldn't update it, especially mix up rebase and merge.

**Comment by karuppiah7890** [comment]
> @karuppiah7890 it's the PR author.

I think we need some sort of `Co-authored-by: ` and put my name in co-author but your name should be in the commit as you are working on it. Or let's create a separate PR

**Comment by karuppiah7890** [comment]
> Short answer, you shouldn't update it, especially mix up rebase and merge.

Cool !

**Comment by tisonkun** [comment]
> I think we need some sort of `Co-authored-by: ` and put my name in co-author but your name should be in the commit as you are working on it. Or let's create a separate PR

So far, the commit message is handled by @ti-chi-bot . We're discussing to change the style https://internals.tidb.io/t/topic/361 . Maybe implement later - I don't have a concrete schedule, though.

**Comment by karuppiah7890** [comment]
Cool ! But please ensure your name is on the commit as you are working on it. It would be nice to have both names with mine as co-author

File: ddl/failtest/fail_db_serial_test.go [hunk 1]
```diff
@@ -340,103 +317,125 @@ func (s *testFailDBSuite) TestGenGlobalIDFail(c *C) {
...
-	defer tk.MustExec(fmt.Sprintf("set @@global.tidb_ddl_reorg_worker_cnt=%d", originDDLAddIndexWorkerCnt))
-
-	if !s.IsCommonHandle { // only enable failpoint once
-		c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/checkBackfillWorkerNum", `return(true)`), IsNil)
-		defer func() {
-			c.Assert(failpoint.Disable("github.com/pingcap/tidb/ddl/checkBackfillWorkerNum"), IsNil)
-		}()
+	tests := []struct {
+		name                   string
+		checkBackfillWorkerNum bool
```
Comment (tangenta) [inline ddl/failtest/fail_db_serial_test.go:326]: It seems this field is not used in the test, can we remove it?

Comment (tisonkun) [reply ddl/failtest/fail_db_serial_test.go]: Nice catch. It is an intermediate variable. Removed.

**Comment by tisonkun** [comment]
> Cool ! But please ensure your name is on the commit as you are working on it. It would be nice to have both names with mine as co-author

I hope so. However, @ti-chi-bot hijacks the commit message. I created an issue https://github.com/ti-community-infra/tichi/issues/733 to investigate whether we can properly respect multiple authors contribution.

**Review by tangenta — APPROVED** [review]

**Pull request merged**